rhasspynlu/jsgf.py

```diff
@@ -478,20 +481,19 @@ def parse_expression(
                 tag.tag_text = text[current_index + 1 : next_index - 1]
 
                 # Handle substitution/converter(s)
-                if ":" in tag.tag_text or "!" in tag.tag_text:
-                    if "!" in tag.tag_text:
-                        # Word with converter(s)
-                        # e.g., twenty:20!int
-                        parts = tag.tag_text.split("!")
-                        tag.tag_text = parts[0]
-                        tag.converters = parts[1:]
-
-                    if ":" in tag.tag_text:
-                        # Word with substitution
-                        # e.g., twenty:20
-                        lhs, rhs = tag.tag_text.split(":", maxsplit=1)
-                        tag.tag_text = lhs
-                        tag.substitution = Substitutable.parse_substitution(rhs)
```

Comment:
This did the same in check to find the character twice with searching a character being a O(n) operation.

Reply:
I think this code evolved over multiple iterations, and I didn't realize the check was being done twice.